torchrec_dlrm/dlrm_main.py

```diff
@@ -391,98 +355,43 @@ def _train(
-            will be run within an epoch.
-        limit_train_batches (Optional[int]): Number of train batches.
-        limit_val_batches (Optional[int]): Number of validation batches.
-
-
+        lr_scheduler (LRPolicyScheduler): Learning rate scheduler.
+        print_lr (bool): Whether to print the learning rate every training step.
+        validation_freq (Optional[int]): The number of training steps between validation runs within an epoch.
+        limit_train_batches (Optional[int]): Limits the training set to the first `limit_train_batches` batches.
+        limit_val_batches (Optional[int]): Limits the validation set to the first `limit_val_batches` batches.
 
     Returns:
         None.
     """
-    train_pipeline._model.train()
-
-    # Because TrainPipelineSparseDist buffer batches internally, we load in
-    # TRAIN_PIPELINE_STAGES - 1 batches from the next_iterator into the buffers so that
-    # when train_val_test switches to the next phase, train_pipeline will start
-    # producing results for the TRAIN_PIPELINE_STAGES - 1 buffered batches (as opposed
-    # to the last TRAIN_PIPELINE_STAGES - 1 batches from iterator).
-    combined_iterator = itertools.chain(
-        iterator
-        if limit_train_batches is None
-        else itertools.islice(iterator, limit_train_batches - (epoch > 0)*(TRAIN_PIPELINE_STAGES - 1)),
-        itertools.islice(next_iterator, TRAIN_PIPELINE_STAGES - 1),
-    )
-    samples_per_trainer_across_epochs = TOTAL_TRAINING_SAMPLES / dist.get_world_size() * epochs
-    samples_per_trainer = TOTAL_TRAINING_SAMPLES / dist.get_world_size()
-    num_batches = samples_per_trainer / batch_size
-    if limit_train_batches is not None:
-        num_batches = min(limit_train_batches, num_batches)
+    pipeline._model.train()
+
+    iterator = itertools.islice(iter(train_dataloader), limit_train_batches)
+    # Filler batches are appended to the end of the iterator to keep the pipeline active while the
+    # last few remaining batches are still in progress awaiting results.
+    two_filler_batches = itertools.islice(iter(train_dataloader), TRAIN_PIPELINE_STAGES - 1)
+    iterator = itertools.chain(iterator, two_filler_batches)
+
     is_rank_zero = dist.get_rank() == 0
     if is_rank_zero:
-        pbar = tqdm(iter(int, 1), desc=f"Epoch {epoch}", disable=False)
+        pbar = tqdm(iter(int, 1), desc=f"Epoch {epoch}", total=len(train_dataloader), disable=False)
     for it in itertools.count():
         try:
-            if is_rank_zero:
-                if print_lr:
-                    for i, g in enumerate(train_pipeline._optimizer.param_groups):
-                        print(f"lr: {it} {i} {g['lr']:.6f}")
-            train_pipeline.progress(combined_iterator)
+            if validation_freq and it > 0 and it % validation_freq == 0:
+                _evaluate(limit_val_batches, new_pipeline(pipeline), val_dataloader, "val")
```

Comment:
do we need to create a new_pipeline each iteration? why not have one eval pipeline? (and just pass in new iter of the dataloader)?

Reply:
If we used the same pipeline for intra-epoch validation, it complicates ordering batches correctly.  When StopIteration is reached in _evaluate, `two_filler_batches` are in the pipeline.  The these two batches are the first two batches of the validation set. That means subsequent intra-epoch validation runs would need to slice off the first two batches from the validation iterator since the first two batches are already in the pipeline.  Additionally, args.limit_val_batches will need to be subtracted by 2 (the pipeline length - 1) when used to limit the iterator.    These differences in behavior between the first run and subsequent runs means there would need to be conditionals expressing if epoch > do this, otherwise do that.  The above changes would increase and complicate the code.  We aren't in a hurry, so we can think about it.  

One note. Under the hood, PyTorch uses already-allocated GPU memory for the intra-validation pipeline buffer because [`with torch.no_grad()` reduces memory consumption](https://pytorch.org/docs/stable/generated/torch.no_grad.html). The reduced memory consumption translates into available already-allocated GPU memory for buffered batches of newly spawned eval-only pipelines.

Reply:
> do we need to create a new_pipeline each iteration? why not have one eval pipeline? (and just pass in new iter of the dataloader)?

I'll scratch my first paragraph above regarding reusing the same pipeline for validation.   It can be done cleanly if we add `pipeline._connected = False` at the start of `def _evaluate`.   This way, we could use the same pipeline for all non-training runs, which means validation and test runs.

Reply:
However, the underscore in _connected is the PEP-8 internal use indicator. That generally means not to tamper with such variables. @colin2328 what are you thinking?